book/src/proposals/program-keys-and-signatures.md

```diff
@@ -0,0 +1,105 @@
+# Program Keys and Signatures
+
+## Problem
+
+Programs cannot generate their own signatures in `process_instruction`
+as defined in the [Cross-Program Invocations](cross-program-invocation.md)
+design.
+
+Lack of programmatic signature generation limits the kinds of programs
+that can be implemented in Solana.  For example, a program cannot take
+ownership of a TokenAccount and later in a different transaction transfer
+the ownership based on the state of another program.  If two users want
+to make a wager in tokens on the outcome of a game in Solana, they must
+transfer tokens to some intermediary that will honor their agreement.
+Currently there is no way to implement this intermediary as a program
+in Solana.
+
+This capability is necessary for many DeFi applications, since they
+require assets to be transferred to an escrow agent until some event
+occurs that determines the new owner.
+
+* Decentralized Exchanges that transfer assets between matching bid and
+ask orders.
+
+* Auctions that transfer assets to the winner.
+
+* Games or prediction markets that collect and redistribute prizes to
+the winners.
+
+## Proposed Solution
+
+The goal of this design is to allow programs to generate deterministic
+program specific addresses and programmatically set
+`KeyedAccount.is_signer` for those addresses when `process_instruction()`
+is invoked.
+
+### Deterministic Pubkey Addresses for Programs
+
+Deterministic pubkey addresses for programs follow a similar derivation
```

Comment:
This is where you should introduce `CreateAddressWithSeed` and highlight that subtle difference, the word `Address`, not `Account` and that unlike everything else in that module, returns a `Pubkey` and not an `Instruction`.

Reply:
hmm, what do you imagine is CreateAddressWithSeed, a new system instruction?  That's not what's being proposed here.

Reply:
You wrote `use system_instruction::create_address_with_seed`.

Reply:
That’s a utility function defined in that module for generating the address that’s used in CreateAccountWithSeed

Reply:
Your confusion is inspiring :). There is an alternative design that i think will be much cleaner.